README.md
![alt tag](./assets/nipplejs.png)
> A vanilla virtual joystick for touch capable interfaces


[![npm](https://img.shields.io/npm/v/nipplejs.svg)](https://npmjs.org/package/nipplejs)
[![npm](https://img.shields.io/npm/dm/nipplejs.svg)](https://npmjs.org/package/nipplejs)

# Table Of Contents
<details>

<!-- toc -->

- [Install](#install)
- [Demo](#demo)
- [Usage](#usage)
- [Options](#options)
  * [`options.zone` defaults to 'body'](#optionszone-defaults-to-body)
  * [`options.color` defaults to 'white'](#optionscolor-defaults-to-white)
  * [`options.size` defaults to 100](#optionssize-defaults-to-100)
  * [`options.threshold` defaults to 0.1](#optionsthreshold-defaults-to-01)
  * [`options.fadeTime` defaults to 250](#optionsfadetime-defaults-to-250)
  * [`options.multitouch` defaults to false](#optionsmultitouch-defaults-to-false)
  * [`options.maxNumberOfNipples` defaults to 1](#optionsmaxnumberofnipples-defaults-to-1)
  * [`options.dataOnly` defaults to false](#optionsdataonly-defaults-to-false)
  * [`options.position` defaults to `{top: 0, left: 0}`](#optionsposition-defaults-to-top-0-left-0)
  * [`options.mode` defaults to 'dynamic'.](#optionsmode-defaults-to-dynamic)
    + [`'dynamic'`](#dynamic)
    + [`'semi'`](#semi)
    + [`'static'`](#static)
  * [`options.restJoystick` defaults to true](#optionsrestjoystick-defaults-to-true)
  * [`options.restOpacity` defaults to 0.5](#optionsrestopacity-defaults-to-05)
  * [`options.catchDistance` defaults to 200](#optionscatchdistance-defaults-to-200)
  * [`options.lockX` defaults to false](#optionslockx-defaults-to-false)
  * [`options.lockY` defaults to false](#optionslocky-defaults-to-false)
  * [`options.shape` defaults to 'circle'](#optionsshape-defaults-to-circle)
    + [`'circle'`](#circle)
    + [`'square'`](#square)
  * [`options.dynamicPage` defaults to false](#optionsdynamicpage-defaults-to-false)
  * [`options.follow` defaults to false](#optionsfollow-defaults-to-false)
- [API](#api)
  * [NippleJS instance (manager)](#nipplejs-instance-manager)
    + [`manager.on(type, handler)`](#managerontype-handler)
    + [`manager.off([type, handler])`](#managerofftype-handler)
    + [`manager.get(identifier)`](#managergetidentifier)
    + [`manager.destroy()`](#managerdestroy)
    + [`manager.ids`](#managerids)
    + [`manager.id`](#managerid)
  * [nipple instance (joystick)](#nipple-instance-joystick)
  * [`joystick.on`, `joystick.off`](#joystickon-joystickoff)
  * [`joystick.el`](#joystickel)
  * [`joystick.show([cb])`](#joystickshowcb)
  * [`joystick.hide([cb])`](#joystickhidecb)
  * [`joystick.add()`](#joystickadd)
  * [`joystick.remove()`](#joystickremove)
  * [`joystick.destroy()`](#joystickdestroy)
  * [`joystick.setPosition(cb, { x, y })`](#joysticksetpositioncb--x-y-)
  * [`joystick.identifier`](#joystickidentifier)
  * [`joystick.trigger(type [, data])`](#joysticktriggertype--data)
  * [`joystick.position`](#joystickposition)
  * [`joystick.frontPosition`](#joystickfrontposition)
  * [`joystick.ui`](#joystickui)
- [Events](#events)
  * [manager only](#manager-only)
    + [`added`](#added)
    + [`removed`](#removed)
  * [manager and joysticks](#manager-and-joysticks)
    + [`start`](#start)
    + [`end`](#end)
    + [`move`](#move)
    + [`dir`](#dir)
    + [`plain`](#plain)
    + [`shown`](#shown)
    + [`hidden`](#hidden)
    + [`destroyed`](#destroyed)
    + [`pressure`](#pressure)
- [Contributing](#contributing)

<!-- tocstop -->

</details>

## Install

```bash
npm install nipplejs --save
```

----

## Demo
Check out the [demo here](http://yoannmoinet.github.io/nipplejs/#demo).

----

## Usage

Import it the way you want into your project :

```javascript
// CommonJS
var manager = require('nipplejs').create(options);
```

```javascript
// AMD
define(['nipplejs'], function (nipplejs) {
    var manager = nipplejs.create(options);
});
```

```javascript
// Module
import nipplejs from 'nipplejs';
```

```html
<!-- Global -->
<script src="./nipplejs.js"></script>
<script>
    var manager = nipplejs.create(options);
</script>
```

**:warning: NB :warning:** Your joystick's container **has** to have its CSS `position` property set, either `absolute`, `relative`, `static`, ....

----

## Options
You can configure your joystick in different ways :

```javascript
var options = {
    zone: Element,                  // active zone
    color: String,
    size: Integer,
    threshold: Float,               // before triggering a directional event
    fadeTime: Integer,              // transition time
    multitouch: Boolean,
    maxNumberOfNipples: Number,     // when multitouch, what is too many?
    dataOnly: Boolean,              // no dom element whatsoever
    position: Object,               // preset position for 'static' mode
    mode: String,                   // 'dynamic', 'static' or 'semi'
    restJoystick: Boolean,
    restOpacity: Number,            // opacity when not 'dynamic' and rested
    lockX: Boolean,                 // only move on the X axis
    lockY: Boolean,                 // only move on the Y axis
    catchDistance: Number,          // distance to recycle previous joystick in
                                    // 'semi' mode
    shape: String,                  // 'circle' or 'square'
    dynamicPage: Boolean,           // Enable if the page has dynamically visible elements
    follow: Boolean,                // Makes the joystick follow the thumbstick
};
```

All options are optional :sunglasses:.

### `options.zone` defaults to 'body'
The dom element in which all your joysticks will be injected.

```html
<div id="zone_joystick"></div>

<script type="text/javascript" src="./nipplejs.js"></script>
<script type="text/javascript">
    var options = {
        zone: document.getElementById('zone_joystick'),
    };
    var manager = nipplejs.create(options);
</script>
```

This zone also serve as the mouse/touch events handler.

It represents the zone where all your joysticks will be active.

### `options.color` defaults to 'white'
The background color of your joystick's elements.

Can be any valid CSS color.

### `options.size` defaults to 100
The size in pixel of the outer circle.

The inner circle is 50% of this size.

### `options.threshold` defaults to 0.1
This is the strength needed to trigger a directional event.

Basically, the center is 0 and the outer is 1.

You need to at least go to 0.1 to trigger a directional event.

### `options.fadeTime` defaults to 250
The time it takes for joystick to fade-out and fade-in when activated or de-activated.

### `options.multitouch` defaults to false
Enable the multitouch capabilities.

If, for reasons, you need to have multiple nipples in the same zone.

Otherwise, it will only get one, and all new touches won't do a thing.

Please note that multitouch is off when in `static` or `semi` modes.

### `options.maxNumberOfNipples` defaults to 1
If you need to, you can also control the maximum number of instances that could be created.

Obviously in a multitouch configuration.

### `options.dataOnly` defaults to false
The library won't draw anything in the DOM and will only trigger events with data.

### `options.position` defaults to `{top: 0, left: 0}`
An object that will determine the position of a `static` mode.

You can pass any of the four `top`, `right`, `bottom` and `left`.

They will be applied as any css property.

Ex :
- `{top: '50px', left: '50px'}`
- `{left: '10%', bottom: '10%'}`

### `options.mode` defaults to 'dynamic'.
Three modes are possible :

#### `'dynamic'`
- a new joystick is created at each new touch.
- the joystick gets destroyed when released.
- **can** be multitouch.

#### `'semi'`
- new joystick is created at each new touch farther than `options.catchDistance` of any previously created joystick.
- the joystick is faded-out when released but not destroyed.
- when touch is made **inside** the `options.catchDistance` a new direction is triggered immediately.
- when touch is made **outside** the `options.catchDistance` the previous joystick is destroyed and a new one is created.
- **cannot** be multitouch.

#### `'static'`
- a joystick is positioned immediately at `options.position`.
- one joystick per zone.
- each new touch triggers a new direction.
- **cannot** be multitouch.

### `options.restJoystick` defaults to true
Reset the joystick's position when it enters the rest state.

### `options.restOpacity` defaults to 0.5
The opacity to apply when the joystick is in a rest position.

### `options.catchDistance` defaults to 200
This is only useful in the `semi` mode, and determine at which distance we recycle the previous joystick.

At 200 (px), if you press the zone into a rayon of 200px around the previously displayed joystick,
it will act as a `static` one.

### `options.lockX` defaults to false
Locks joystick's movement to the x (horizontal) axis

### `options.lockY` defaults to false
Locks joystick's movement to the y (vertical) axis

### `options.shape` defaults to 'circle'
The shape of region within which joystick can move.

#### `'circle'`
Creates circle region for joystick movement

#### `'square'`
Creates square region for joystick movement

### `options.dynamicPage` defaults to false
Enable if the page has dynamically visible elements such as for Vue, React, Angular or simply some CSS hiding or showing some DOM.

### `options.follow` defaults to false
Makes the joystick follow the thumbstick when it reaches the border.

----

## API

### NippleJS instance (manager)

Your manager has the following signature :

```javascript
{
    on: Function,                       // handle internal event
    off: Function,                      // un-handle internal event
    get: Function,                      // get a specific joystick
    destroy: Function,                  // destroy everything
    ids: Array                          // array of assigned ids
    id: Number                          // id of the manager
    options: {
        zone: Element,                  // reactive zone
        multitouch: Boolean,
        maxNumberOfNipples: Number,
        mode: String,
        position: Object,
        catchDistance: Number,
        size: Number,
        threshold: Number,
        color: String,
        fadeTime: Number,
        dataOnly: Boolean,
        restJoystick: Boolean,
        restOpacity: Number
    }
}
```

#### `manager.on(type, handler)`

If you wish to listen to internal events like :

```javascript
manager.on('event#1 event#2', function (evt, data) {
    // Do something.
});
```

Note that you can listen to multiple events at once by separating
them either with a space or a comma (or both, I don't care).

You can also separate events in an array.
```javascript
manager.on(['event#1', 'event#2'], function (evt, data) {
    // Do something.
});
```

#### `manager.off([type, handler])`

To remove an event handler :

```javascript
manager.off('event', handler);
```

If you call off without arguments, all handlers will be removed.

If you don't specify the handler but just a type, all handlers for that type will be removed.

#### `manager.get(identifier)`

A helper to get an instance via its identifier.

```javascript
// Will return the nipple instantiated by the touch identified by 0
manager.get(0);
```

#### `manager.destroy()`

Gently remove all nipples from the DOM and unbind all events.

```javascript
manager.destroy();
```

#### `manager.ids`

The array of nipples' ids under this manager.

#### `manager.id`

The incremented id of this manager.

### nipple instance (joystick)

Each joystick has the following signature :

```javascript
{
    on: Function,
    off: Function,
    el: Element,
    show: Function,         // fade-in
    hide: Function,         // fade-out
    add: Function,          // inject into dom
    remove: Function,       // remove from dom
    destroy: Function,
    setPosition: Function,
    identifier: Number,
    trigger: Function,
    position: {             // position of the center
        x: Number,
        y: Number
    },
    frontPosition: {        // position of the front part
        x: Number,
        y: Number
    },
    ui: {
        el: Element,
        front: Element,
        back: Element
    },
    options: {
        color: String,
        size: Number,
        threshold: Number,
        fadeTime: Number
    }
}
```

### `joystick.on`, `joystick.off`

The same as the manager.

### `joystick.el`

Dom element in which the joystick gets created.

```html
<div class="nipple">
    <div class="front"></div>
    <div class="back"></div>
</div>
```

### `joystick.show([cb])`

Will show the joystick at the last known place.

You can pass a callback that will be executed at the end of the fade-in animation.

### `joystick.hide([cb])`

Will fade-out the joystick.

You can pass a callback that will be executed at the end of the fade-out animation.

### `joystick.add()`

Add the joystick's element to the dom.

### `joystick.remove()`

Remove the joystick's element from the dom.

### `joystick.destroy()`

Gently remove this nipple from the DOM and unbind all related events.

### `joystick.setPosition(cb, { x, y })`

Set the joystick to the specified position, where x and y are distances away from the center in pixels. This does not trigger joystick events.

### `joystick.identifier`

Returns the unique identifier of the joystick.

Tied to its touch's identifier.

### `joystick.trigger(type [, data])`

Trigger an internal event from the joystick.

The same as `on` you can trigger multiple events at the same time.

### `joystick.position`

The absolute position of the center of the joystick.

### `joystick.frontPosition`

The absolute position of the back part of the joystick's ui.

### `joystick.ui`

The object that store its ui elements

```html
{
    el: <div class="nipple"></div>
    back: <div class="back"></div>
    front: <div class="front"></div>
}
```

----

## Events

You can listen events both on the manager and all the joysticks.

But some of them are specific to its instance.

If you need to listen to each joystick, for example, you can :

```javascript
manager.on('added', function (evt, nipple) {
    nipple.on('start move end dir plain', function (evt) {
        // DO EVERYTHING
    });
}).on('removed', function (evt, nipple) {
    nipple.off('start move end dir plain');
});
```

### manager only

#### `added`

A joystick just got added.

Will pass the instance alongside the event.

#### `removed`

A joystick just got removed.

Fired at the end of the fade-out animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

### manager and joysticks

Other events are available on both the manager and joysticks.

When listening on the manager,
you can also target **a joystick in particular** by prefixing
the event with its identifier, **`0:start`** for example.

Else you'll get all events from all the joysticks.

#### `start`

A joystick is activated. (the user pressed on the active zone)

Will pass the instance alongside the event.

#### `end`

A joystick is de-activated. (the user released the active zone)

Will pass the instance alongside the event.

#### `move`

A joystick is moved.

Comes with data :

```javascript
{
    identifier: 0,              // the identifier of the touch/mouse that triggered it
    position: {                 // absolute position of the center in pixels
        x: 125,
        y: 95
    },
    force: 0.2,                 // strength in %
    distance: 25.4,             // distance from center in pixels
    pressure: 0.1,              // the pressure applied by the touch
    angle: {
        radian: 1.5707963268,   // angle in radian
        degree: 90
    },
    vector: {                   // force unit vector
      x: 0.508,
      y: 3.110602869834277e-17
    },
    raw: {                      // note: angle is the same, beyond the 50 pixel limit
        distance: 25.4,         // distance which continues beyond the 50 pixel limit
        position: {             // position of the finger/mouse in pixels, beyond joystick limits
            x: 125,
            y: 95
        }
    },
    instance: Nipple            // the nipple instance that triggered the event
}
```

#### `dir`

When a direction is reached after the threshold.

Direction are split with a 45° angle.

```javascript
//     \  UP /
//      \   /
// LEFT       RIGHT
//      /   \
//     /DOWN \
```

You can also listen to specific direction like :

- `dir:up`
- `dir:down`
- `dir:right`
- `dir:left`

In this configuration only one direction is triggered at a time.

#### `plain`

When a plain direction is reached after the threshold.

Plain directions are split with a 90° angle.

```javascript
//       UP               |
//     ------        LEFT | RIGHT
//      DOWN              |
```

You can also listen to specific plain direction like :

- `plain:up`
- `plain:down`
- `plain:right`
- `plain:left`

In this configuration two directions can be triggered at a time,
because the user could be both `up` and `left` for example.

#### `shown`

Is triggered at the end of the fade-in animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

#### `hidden`

Is triggered at the end of the fade-out animation.

Will pass the instance alongside the event.

Won't be trigger in a `dataOnly` configuration.

#### `destroyed`

Is triggered at the end of destroy.

Will pass the instance alongside the event.

#### `pressure`

> MBP's [**Force Touch**](http://www.apple.com/macbook-pro/features-retina/#interact), iOS's [**3D Touch**](http://www.apple.com/iphone-6s/3d-touch/), Microsoft's [**pressure**](https://msdn.microsoft.com/en-us/library/hh772360%28v=vs.85%29.aspx) or MDN's [**force**](https://developer.mozilla.org/en-US/docs/Web/API/Touch/force)

Is triggered when the pressure on the joystick is changed.

The value, between 0 and 1, is sent back alongside the event.

----

## Contributing
You can follow [this document](./CONTRIBUTING.md) to help you get started.


CHANGELOG.md
## [0.10.1](https://github.com/yoannmoinet/nipplejs/compare/v0.10.0...v0.10.1) (2023-03-06)


### Bug Fixes

* **manager:** add `ids` & `id` to types ([#207](https://github.com/yoannmoinet/nipplejs/issues/207)) ([6b7b6dc](https://github.com/yoannmoinet/nipplejs/commit/6b7b6dceb7e876fbc0f3c12b8e7e9069353079cf))



# [0.10.0](https://github.com/yoannmoinet/nipplejs/compare/v0.9.1...v0.10.0) (2022-10-10)


### Bug Fixes

* missing type for `setPosition` ([#194](https://github.com/yoannmoinet/nipplejs/issues/194)) ([3a2e495](https://github.com/yoannmoinet/nipplejs/commit/3a2e495873ae786d1502e759b2dcaa17a3bc0835))
* use transform translate for nipple front ([#187](https://github.com/yoannmoinet/nipplejs/issues/187)) ([52e5bd2](https://github.com/yoannmoinet/nipplejs/commit/52e5bd2c5498328478ab2e5db2c65714f5a93f53))



## [0.9.1](https://github.com/yoannmoinet/nipplejs/compare/v0.9.0...v0.9.1) (2022-03-26)


### Bug Fixes

* **zone:** handle offset when parent is flex ([#182](https://github.com/yoannmoinet/nipplejs/issues/182)) ([ef9ec5a](https://github.com/yoannmoinet/nipplejs/commit/ef9ec5a7ca2680ad233b4d7fccc71f6c025aca50))


### Features

* **improvement:** restJoystick now accepts an object ([#162](https://github.com/yoannmoinet/nipplejs/issues/162)) ([ce3a67d](https://github.com/yoannmoinet/nipplejs/commit/ce3a67dbfa7b64c5de9ff4a74c008d46e42f5194))



# [0.9.0](https://github.com/yoannmoinet/nipplejs/compare/v0.8.7...v0.9.0) (2021-02-24)


### Features

* **follow:** Add option to follow thumbstick ([#160](https://github.com/yoannmoinet/nipplejs/issues/160)) ([5c4c002](https://github.com/yoannmoinet/nipplejs/commit/5c4c0027b73fa25b5410d478199844a5277c1e76))



## [0.8.7](https://github.com/yoannmoinet/nipplejs/compare/v0.8.6...v0.8.7) (2020-09-06)



## [0.8.6](https://github.com/yoannmoinet/nipplejs/compare/v0.8.3...v0.8.6) (2020-09-06)


### Bug Fixes

* add missing type for `dynamicPage` ([cc97650](https://github.com/yoannmoinet/nipplejs/commit/cc97650db299b0a3b5db4da0a98a561825904dc7))


### Features

* add an options to use with dynamic pages ([8258501](https://github.com/yoannmoinet/nipplejs/commit/8258501721b9972a9b2206a9a67224a200cdf83b))
* Add shape option to joystick ([#107](https://github.com/yoannmoinet/nipplejs/issues/107)) ([5a8f8ba](https://github.com/yoannmoinet/nipplejs/commit/5a8f8bacea3776630a270a25dbbd760c9451bb55))
* add unit vector representation to move event output ([a4496fe](https://github.com/yoannmoinet/nipplejs/commit/a4496fee62ca59f5097e972faf696e4e17a33829))



## [0.8.3](https://github.com/yoannmoinet/nipplejs/compare/v0.8.2...v0.8.3) (2019-08-24)


### Bug Fixes

* better lockX and lockY mecanism ([8355fb8](https://github.com/yoannmoinet/nipplejs/commit/8355fb8dd884541e2c1c4f5ad0f69ee2674be1ca)), closes [#111](https://github.com/yoannmoinet/nipplejs/issues/111)
* released outside will be considered inside ([9a8a7b4](https://github.com/yoannmoinet/nipplejs/commit/9a8a7b4da42e9342621560a0588f580a31dbb419)), closes [#109](https://github.com/yoannmoinet/nipplejs/issues/109)
* reset direction when below the threshold ([6f41b38](https://github.com/yoannmoinet/nipplejs/commit/6f41b38e1c6b06b6bf0fe86ba4d5a4668689dc0d)), closes [#110](https://github.com/yoannmoinet/nipplejs/issues/110)



## [0.8.2](https://github.com/yoannmoinet/nipplejs/compare/v0.8.1...v0.8.2) (2019-05-08)


### Bug Fixes

* allow older typescript ([c164597](https://github.com/yoannmoinet/nipplejs/commit/c164597d5cc8d18f4ed2553a76554b1995d78e97))
* joystick in 'dynamic' mode gets stuck on iOS ([94bb784](https://github.com/yoannmoinet/nipplejs/commit/94bb784fecc5eeb5a7fcf03675fa074046fd9663)), closes [#94](https://github.com/yoannmoinet/nipplejs/issues/94)


### Features

* add raw data on move for data-only usage ([c818fb2](https://github.com/yoannmoinet/nipplejs/commit/c818fb2daffb7333e919f427ae53cb5ee9fce047)), closes [#54](https://github.com/yoannmoinet/nipplejs/issues/54)



## [0.8.1](https://github.com/yoannmoinet/nipplejs/compare/v0.7.3...v0.8.1) (2019-02-17)


### Features

* upgrade dev experience with webpack and es6 ([cc1e824](https://github.com/yoannmoinet/nipplejs/commit/cc1e824c32527a2c85f8b053d7134317b6e93563))



## [0.7.3](https://github.com/yoannmoinet/nipplejs/compare/v0.7.2...v0.7.3) (2018-12-11)



## [0.7.2](https://github.com/yoannmoinet/nipplejs/compare/v0.7.1...v0.7.2) (2018-12-11)


### Features

* Add TypeScript Definitions ([#85](https://github.com/yoannmoinet/nipplejs/issues/85)) ([0b1f4ae](https://github.com/yoannmoinet/nipplejs/commit/0b1f4ae0d54b41ef1a9c1feef4f820b8a85b8519)), closes [#21](https://github.com/yoannmoinet/nipplejs/issues/21)



## [0.7.1](https://github.com/yoannmoinet/nipplejs/compare/v0.7.0...v0.7.1) (2018-06-06)



# [0.7.0](https://github.com/yoannmoinet/nipplejs/compare/v0.6.8...v0.7.0) (2018-06-06)


### Bug Fixes

* register touchcancel and pointercancel events ([2c541a6](https://github.com/yoannmoinet/nipplejs/commit/2c541a60427a92b3b26582ed781fac99399b41f5))


### Features

* **joystick:** add lockX and lockY options ([6baba97](https://github.com/yoannmoinet/nipplejs/commit/6baba9707999f77336efd3d87f2d3a04a87d5d80))



## [0.6.8](https://github.com/yoannmoinet/nipplejs/compare/v0.6.7...v0.6.8) (2017-12-19)


### Bug Fixes

* add touchcancel handling as touchend for events coming from outside ([ebc501e](https://github.com/yoannmoinet/nipplejs/commit/ebc501e2760af6dd048cac4dbd51d57e0e160a61)), closes [#61](https://github.com/yoannmoinet/nipplejs/issues/61) [#57](https://github.com/yoannmoinet/nipplejs/issues/57) [#33](https://github.com/yoannmoinet/nipplejs/issues/33) [#31](https://github.com/yoannmoinet/nipplejs/issues/31) [#30](https://github.com/yoannmoinet/nipplejs/issues/30)
* update manager with changing identifier ([45ec1d7](https://github.com/yoannmoinet/nipplejs/commit/45ec1d7f23064b8e2096e3ab1120a69a5147ce9f)), closes [#64](https://github.com/yoannmoinet/nipplejs/issues/64)


### Features


types/index.d.ts
export interface JoystickManagerOptions {
    /**
     * Defaults to `'body'`
     * The dom element in which all your joysticks will be injected.
     *
     * This zone also serve as the mouse/touch events handler.
     *
     * It represents the zone where all your joysticks will be active.
     */
    zone?: HTMLElement;

    /**
     * Defaults to `'white'`
     * The background color of your joystick’s elements.
     *
     * Can be any valid CSS color.
     */
    color?: string;

    /**
     * Defaults to `100`
     * The size in pixel of the outer circle.
     *
     * The inner circle is 50% of this size.
     */
    size?: number;

    /**
     * This is the strength needed to trigger a directional event.
     *
     * Basically, the center is 0 and the outer is 1.
     *
     * You need to at least go to 0.1 to trigger a directional event.
     */
    threshold?: number;

    /**
     * Defaults to `250`
     *
     * The time it takes for joystick to fade-out and fade-in when activated or de-activated.
     */
    fadeTime?: number;

    /**
     * Defaults to `false`
     *
     * Enable the multitouch capabilities.
     *
     * If, for reasons, you need to have multiple nipples into the same zone.
     *
     * Otherwise it will only get one, and all new touches won’t do a thing.
     *
     * Please note that multitouch is off when in static or semi modes.
     */
    multitouch?: boolean;

    /**
     * Defaults to `1`
     *
     * If you need to, you can also control the maximum number of instance that could be created.
     *
     * Obviously in a multitouch configuration.
     */
    maxNumberOfNipples?: number;

    /**
     * Defaults to `false`
     *
     * The library won’t draw anything in the DOM and will only trigger events with data.
     */
    dataOnly?: boolean;

    /**
     * Defaults to `{top: 0, left: 0}`
     *
     * An object that will determine the position of a static mode.
     *
     * You can pass any of the four top, right, bottom and left.
     *
     * They will be applied as any css property.
     */
    position?: {
        top?: string;
        right?: string;
        bottom?: string;
        left?: string;
    };

    /**
     * Behavioral mode for the joystick
     *
     * ### 'dynamic':
     * a new joystick is created at each new touch.
     * the joystick gets destroyed when released.
     * can be multitouch.
     *
     * ### 'semi':
     * new joystick is created at each new touch farther than options.catchDistance of any previously created joystick.
     * the joystick is faded-out when released but not destroyed.
     * when touch is made inside the options.catchDistance a new direction is triggered immediately.
     * when touch is made outside the options.catchDistance the previous joystick is destroyed and a new one is created.
     * cannot be multitouch.
     *
     * ### 'static':
     * a joystick is positioned immediately at options.position.
     * one joystick per zone.
     * each new touch triggers a new direction.
     * cannot be multitouch.
     */
    mode?: 'dynamic' | 'semi' | 'static';

    /**
     * Defaults to `true`
     *
     * Reset the joystick’s position when it enters the rest state.
     */
    restJoystick?: boolean | RestJoystickOption;

    /**
     * Defaults to `0.5`
     * The opacity to apply when the joystick is in a rest position.
     */
    restOpacity?: number;

    /**
     * Defaults to `200`
     *
     * This is only useful in the semi mode, and determine at which distance we recycle the previous joystick.
     */
    catchDistance?: number;

    /**
     * Defaults to `false`
     *
     * Locks joystick’s movement to the x (horizontal) axis
     */
    lockX?: boolean;

    /**
     * Defaults to `false`
     *
     * Locks joystick’s movement to the y (vertical) axis
     */
    lockY?: boolean;

    /**
     * Defaults to `false`
     *
     * Enable if the page has dynamically visible elements such as for Vue, React, Angular or simply some CSS hiding or
     * showing some DOM.
     */
    dynamicPage?: boolean;

    /**
     * Defaults to `circle`
     *
     * Sets the shape of the joystick
     */
    shape?: 'circle' | 'square';

    /**
     * Defaults to `false`
     *
     * Make the joystick follow the cursor beyond its limits.
     */
    follow?: boolean;
}

export interface RestJoystickOption {
    /**
     * Defaults to `true`
     */
    x?: boolean,
    /**
     * Defaults to `true`
     */
    y?: boolean,
}

export interface Position {
    x: number;
    y: number;
}

export interface Direction {
    angle: 'up' | 'down' | 'right' | 'left';
    x: 'left' | 'right';
    y: 'up' | 'down';
}

export interface JoystickOutputData {
    angle: {
        degree: number;
        radian: number;
    };
    direction: Direction;
    vector: {
        x: number;
        y: number;
    };
    raw: {
      distance: number;
      position: Position;
    };
    distance: number;
    force: number;
    identifier: number;
    instance: Joystick;
    position: Position;
    pressure: number;
}

export type JoystickEventTypes =
    /**
     * A joystick is activated. (the user pressed on the active zone)
     * Will pass the instance alongside the event.
     */
    | 'start'

    /** A joystick is de-activated. (the user released the active zone)
     * Will pass the instance alongside the event.
     */
    | 'end'

    /**
     * A joystick is moved
     */
    | 'move'

    /**
     * When a direction is reached after the threshold.
     *
     * Direction are split with a 45° angle.
     */
    | 'dir'
    | 'dir:up'
    | 'dir:down'
    | 'dir:right'
    | 'dir:left'

    /**
     * When a plain direction is reached after the threshold.
     * Plain directions are split with a 90° angle.
     */
    | 'plain'

    // plain variations
    | 'plain:up'
    | 'plain:down'
    | 'plain:right'
    | 'plain:left'

    /**
     * Is triggered at the end of the fade-in animation.
     * Will pass the instance alongside the event.
     *
     * Won’t be trigger in a dataOnly configuration.
     */
    | 'shown'

    /**
     * Is triggered at the end of the fade-out animation.
     *
     * Will pass the instance alongside the event.
     *
     * Won’t be trigger in a dataOnly configuration.
     */
    | 'hidden'

    /**
     * Is triggered at the end of destroy.
     *
     * Will pass the instance alongside the event.
     */
    | 'destroyed'

    /**
     * MBP’s Force Touch, iOS’s 3D Touch, Microsoft’s pressure or MDN’s force
     *
     *Is triggered when the pressure on the joystick is changed.
     *
     *The value, between 0 and 1, is sent back alongside the event.
     */
    | 'pressure';

export type ManagerOnlyEventTypes =
    /**
     * A joystick just got added.
     *
     * Will pass the instance alongside the event.
     */
    | 'added'

    /**
     * A joystick just got removed.
     *
     * Fired at the end of the fade-out animation.
     *
     * Will pass the instance alongside the event.
     *
     * Won’t be trigger in a dataOnly configuration.
     */
    | 'removed';

export type JoystickManagerEventTypes = JoystickEventTypes | ManagerOnlyEventTypes;

export interface EventData {
    type: JoystickEventTypes | ManagerOnlyEventTypes;
    target: Collection;
}

export class JoystickManager {
    create(options?: JoystickManagerOptions): JoystickManager;

    on(
        type: JoystickManagerEventTypes | JoystickManagerEventTypes[],
        handler: (evt: EventData, data: JoystickOutputData) => void
    ): void;
    off(
        type: JoystickManagerEventTypes | JoystickManagerEventTypes[],
        handler: (evt: EventData, data: JoystickOutputData) => void
    ): void;
    get(identifier: number): Joystick;
    destroy(): void;
    ids: number[];
    id: number;
}

export interface Collection {
    nipples: Joystick[];
    idles: Joystick[];
    actives: Joystick[];
    ids: number[];
    pressureIntervals: {};
    manager: JoystickManager;
    id: number;
    defaults: JoystickManagerOptions;
    parentIsFlex: boolean;
}

export interface Joystick {
    on(
        type: JoystickEventTypes | JoystickEventTypes[],
        handler: (evt: EventData, data: JoystickOutputData) => void
    ): void;
    off(
        type: JoystickEventTypes | JoystickEventTypes[],
        handler: (evt: EventData, data: JoystickOutputData) => void
    ): void;
    el: HTMLElement;
    show(cb?: () => void): void;
    hide(cb?: () => void): void;
    add(): void;
    remove(): void;
    destroy(): void;
    setPosition(cb: (joystick: Joystick) => void, position: Position): void;
    identifier: number;
    trigger(
        type: JoystickEventTypes | JoystickEventTypes[],
        handler: (evt: EventData, data: any) => void
    ): void;
    position: Position;
    frontPosition: Position;
    ui: {
        el: HTMLElement;
        front: HTMLElement;
        back: HTMLElement;
    };
    options: JoystickManagerOptions;
}

/**
 * A JavaScript library for creating vanillaJS virtual joysticks, for touch capable interfaces.
 */
declare module 'nipplejs' {
    /**
     * Create a Joystick manager
     * @param options for creating a manager instance
     * @return manager instance
     */
    function create(options: JoystickManagerOptions): JoystickManager;

    /**
     * Library's root manger instance.
     */
    const factory: JoystickManager;
}
